### PR TITLE
Adding kinematic fit

### PR DESCRIPTION
**Pull request by andersbarzdukas**
**Changes:**
-Added kinematic refit and subsequent variables to picos (not using rochester corrected pT)
-Changed photon requirement to include photon_wp80, mindr > 0.3, and photon_elidx==-1 (overlapping object rejection)
-Added lepton charge requirement and removed ll_charge branch
-Changed skim requirements to require nll>=1
-Added photon_jet_maxdr, PTt variables, and llphoton_dijet_dr
-Added lep* to skim variables
-Changed use_event to be true for data

For kinematic refit changes please look at kingscanyon_v1.addingkinematicfit/ picos on cms11. Some other changes are not included in these test picos

**Possible additions before merge:**
-Adding pixelseed veto to photon reconstruction
-Other changes?

### CODE REVIEW

**Comment by mcoshiro** [comment]
One possible suggestion to consider: perhaps we should only require photon_elidx does not point to a signal electron. The current scheme removes all photons that appear in the pico electron list (that pass pt, eta, dxy, and dz cuts, but not necessarily ID), but this introduces extra complication for synchronization, and possibly a loss of efficiency.

Other than that, the other changes all look good to merge to me.

**Comment by andersbarzdukas** [comment]
I added the change to photon definitions in the most recent push. The definition of a photon should reject photons if the el_sig variable at that particular photon_elidx index is true. If the photon is not in the electron list it should always pass this criteria. 

I also had forgotten to include the most up to date variables file, so that should be good now as well.

File: inc/dilep_producer.hpp
Comment (jaebak) [inline inc/dilep_producer.hpp:1]: Let's remove the "#include "KinZfitter.hpp" and the comments in line 18 to 21. I don't think they are used?

File: src/dilep_producer.cpp
Comment (jaebak) [inline src/dilep_producer.cpp:1]: Again, let's remove the unnecessary code.
For line 3,  line 11 to 15, and line 19 to 23

File: src/dilep_producer.cpp [hunk 1]
```diff
@@ -22,21 +32,30 @@ void DileptonProducer::WriteDileptons(pico_tree &pico,
...
         mu1.SetPtEtaPhiM(pico.out_mu_pt()[imu1], pico.out_mu_eta()[imu1],
                          pico.out_mu_phi()[imu1], 0.10566);
         mu2.SetPtEtaPhiM(pico.out_mu_pt()[imu2], pico.out_mu_eta()[imu2],
                          pico.out_mu_phi()[imu2], 0.10566);
         dimu = mu1 + mu2;
+        
+
         // Dilepton closest to Z mass gets put at the front
-        if(abs(dimu.M() - zmass) < mindm) { 
+        if((abs(dimu.M() - zmass) < mindm)) {
```
Comment (jaebak) [inline src/dilep_producer.cpp:58]: Why did you add in an additional parenthesis?

File: src/dilep_producer.cpp [hunk 2]
```diff
@@ -110,8 +135,10 @@ void DileptonProducer::WriteDileptons(pico_tree &pico,
         pico.out_ll_i2()   .insert(pico.out_ll_i2()   .begin()+shift, iel2);
         pico.out_ll_l1_masserr() .insert(pico.out_ll_l1_masserr() .begin()+shift, dml1);
         pico.out_ll_l2_masserr() .insert(pico.out_ll_l2_masserr() .begin()+shift, dml2);
-        pico.out_ll_charge()     .insert(pico.out_ll_charge()     .begin()+shift, pico.out_el_charge()[iel1]+pico.out_el_charge()[iel2]);
+        //pico.out_ll_charge()     .insert(pico.out_ll_charge()     .begin()+shift, dilep_charge);
         nll++;
       }
+
```
Comment (jaebak) [inline src/dilep_producer.cpp:141]: Line 141 and 142 is not needed.

File: src/zgamma_producer.cpp [hunk 3]
```diff
@@ -2,14 +2,18 @@
...
 
 using namespace std;
 
 ZGammaVarProducer::ZGammaVarProducer(int year_){
     year = year_;
+    kinZfitter = new KinZfitter();
 }
 
 ZGammaVarProducer::~ZGammaVarProducer(){
+    kinZfitter = new KinZfitter();
```
Comment (jaebak) [inline src/zgamma_producer.cpp:15]: This should be delete instead of new.

File: src/zgamma_producer.cpp [hunk 4]
```diff
@@ -32,11 +36,15 @@ void ZGammaVarProducer::WriteZGammaVars(nano_tree &nano, pico_tree &pico, vector
...
+
+  for(size_t ill(0); ill < pico.out_ll_pt().size(); ill++){
     for(size_t igamma(0); igamma < pico.out_photon_pt().size(); igamma++) {
-      TLorentzVector dilep, photon, llg;
+      TLorentzVector dilep, dilep_refit, photon, llg;
       dilep.SetPtEtaPhiM(pico.out_ll_pt()[ill], pico.out_ll_eta()[ill],
                          pico.out_ll_phi()[ill],pico.out_ll_m()[ill]);
+      //dilep_refit.SetPtEtaPhiM(pico.out_ll_refit_pt()[ill], pico.out_ll_refit_eta()[ill],
+      //                         pico.out_ll_refit_phi()[ill],pico.out_ll_refit_m()[ill]);
+
```
Comment (jaebak) [inline src/zgamma_producer.cpp:47]: Let's remove commented code.

File: src/zgamma_producer.cpp [hunk 5]
```diff
@@ -47,6 +55,16 @@ void ZGammaVarProducer::WriteZGammaVars(nano_tree &nano, pico_tree &pico, vector
...
+      //llg_refit = dilep_refit + photon; 
+      //pico.out_llphoton_refit_pt().push_back(llg_refit.Pt());
+      //pico.out_llphoton_refit_eta().push_back(llg_refit.Eta());
+      //pico.out_llphoton_refit_phi().push_back(llg_refit.Phi());
+      //pico.out_llphoton_refit_m().push_back(llg_refit.M());
+      //pico.out_llphoton_refit_dr().push_back(dilep_refit.DeltaR(photon));
+      //pico.out_llphoton_refit_dphi().push_back(dilep_refit.DeltaPhi(photon));
+      //pico.out_llphoton_refit_deta().push_back(fabs(dilep_refit.Eta() - photon.Eta()));
+
```
Comment (jaebak) [inline src/zgamma_producer.cpp:67]: Let's remove commented code.

File: src/zgamma_producer.cpp [hunk 6]
```diff
@@ -56,13 +74,21 @@ void ZGammaVarProducer::WriteZGammaVars(nano_tree &nano, pico_tree &pico, vector
...
         pico.out_photon_zeppenfeld().push_back(abs(photon.Eta() - (j1.Eta() + j2.Eta())/2));
+
         TVector3 g = photon.Vect();
         TVector3 h = llg.Vect();
         TVector3 z = dilep.Vect();
         g.SetZ(0); h.SetZ(0); z.SetZ(0);
-        pico.out_llphoton_pTt2().push_back(h.Cross(z-g).Mag()/h.Mag());
+        pico.out_llphoton_pTt().push_back( fabs(h.Cross(z-g).Pz())/(z-g).Pt());
+        pico.out_llphoton_pTt_an_hig019014().push_back( fabs(h.Cross(z-g).Pz())/h.Mag() );
+        pico.out_llphoton_pTt_oldpicos().push_back(h.Cross(z-g).Mag()/h.Mag());
```
Comment (jaebak) [inline src/zgamma_producer.cpp:91]: I believe the correct way in calculating ptt is,

h.Cross((z-g).Unit()).Mag()

The way that hig19-014 calculated this was "h.Cross(z-g).Mag()/h.Mag()"

Let's not have 3 calculation methods. Let's just have the correct ptt and the ptt calculated by hig19-014.

File: src/zgamma_producer.cpp [hunk 7]
```diff
@@ -195,7 +221,110 @@ void ZGammaVarProducer::WriteZGammaVars(nano_tree &nano, pico_tree &pico, vector
...
+      pico.out_ll_refit_covmat_status() = covmatstatus;
+      pico.out_ll_refit_minnll()        = minnll;
+
+      pico.out_llphoton_refit_pt()   = llg_refit.Pt();
+      pico.out_llphoton_refit_eta()  = llg_refit.Eta();
+      pico.out_llphoton_refit_phi()  = llg_refit.Phi();
+      pico.out_llphoton_refit_m()    = llg_refit.M();
+      pico.out_llphoton_refit_dr()   = ll_refit.DeltaR(photon);
+      pico.out_llphoton_refit_dphi() = ll_refit.DeltaPhi(photon);
+      pico.out_llphoton_refit_deta() = fabs(ll_refit.Eta() - photon.Eta());
```
Comment (jaebak) [inline src/zgamma_producer.cpp:323]: Many event variables are related with the leptons. I wonder if we want to have additional recalculated event variables.
- photon_drmin
- photon_drmax
- llphoton_cosTheta
- llphoton_costheta
- llphoton_psi
- refit_leptons (eta, phi)
- ptt

Also the dijet variables get modified
- llphoton_dijet_dphi
- llphoton_dijet_balance

It would be nice to see how different things become, before doing the production.

File: src/zgamma_producer.cpp (hunk 6 above)
Comment (andersbarzdukas) [reply src/zgamma_producer.cpp:91]: Do we want the hig-019-014 version  of the variable in the AN or the paper.
The paper version is the first of the three there, and the AN version is on line 90.

I dont believe that the h.Cross((z-g).Unit()).Mag() is correct though because h,z,g are 3D vectors where in the paper they only care about the cross product of the pT of the higgs candidate and the Z-g object (see below). The denominator on line 90 should be h.Pt() though.

AN definition: image
Paper: image

File: src/zgamma_producer.cpp (hunk 7 above)
Comment (andersbarzdukas) [reply src/zgamma_producer.cpp:323]: I will check this in a sample file. I dont imagine they will change a lot, but should I update these variables with the refit value or should there be new variables?

File: src/zgamma_producer.cpp (hunk 6 above)
Comment (jaebak) [reply src/zgamma_producer.cpp:91]: Line 88 sets the gamma, higgs and Z to be the pT "vector". So I think "h.Cross((z-g).Unit()).Mag()" is correct.

I don't care that much of what is written in the AN. I think we should have a version that was used in the HIG19014 analysis code, which was the ptt2 variable.

**Comment by andersbarzdukas** [comment]
Ok I believe I have made all the requested changes with the last push. I will look into the other variables and include then either as new variables or change the current ones when we come to a decision on that.

File: src/zgamma_producer.cpp (hunk 7 above)
Comment (mcoshiro) [reply src/zgamma_producer.cpp:310]: Don't know if this is resolved yet, but shouldn't photon_drmin/photon_drmax/lep_eta/lep_phi be unchanged? I was under the impression that the refit only changes lepton pT (and thus things calculated from them including the llγ system angles etc.)

Comment (jaebak) [reply src/zgamma_producer.cpp:310]: That is true. I crossed off them from the list.

Comment (jaebak) [reply src/zgamma_producer.cpp:310]: I think we should have both the refit value and the pre-fit value. The pre-fit value could be useful for synchronization.

File: src/zgamma_producer.cpp [hunk 8]
```diff
@@ -49,21 +54,31 @@ void ZGammaVarProducer::WriteZGammaVars(nano_tree &nano, pico_tree &pico, vector
...
-        pico.out_llphoton_pTt2().push_back(h.Cross(z-g).Mag()/h.Mag());
+
       }
+
+      TVector3 g_pT = photon.Vect();
+      TVector3 h_pT = llg.Vect();
+      TVector3 z_pT = dilep.Vect();
+      g_pT.SetZ(0); h_pT.SetZ(0); z_pT.SetZ(0);
+      pico.out_llphoton_pTt().push_back( h_pT.Cross((z_pT-g_pT).Unit()).Mag() );
+      pico.out_llphoton_pTt_an_hig019014().push_back( h_pT.Unit().Cross(z_pT-g_pT).Mag() );
```
Comment (jaebak) [inline src/zgamma_producer.cpp:80]: Why is "h_pT.Unit().Cross(z_pT-g_pT).Mag()" different with "h.Cross(z-g).Mag()/h.Mag()" that was there before?

Comment (andersbarzdukas) [reply src/zgamma_producer.cpp:80]: This is the same variable with a slightly different way of writing it, no? This way seems like a bit more clear in the difference between pTt and pTt_an_hig019014. I can move both to be written either way, but being consistent may be nice.

**Comment by andersbarzdukas** [comment]
Pull Request passed review. Will merge now
